### PR TITLE
feat: add port config

### PR DESCRIPTION
**Pull request by k-khatri**
**Issue #, if available:**
GG-36859
**Description of changes:**
Add port config so that customers can configure it when the MQTT port for broker is changed
**Why is this change necessary:**
Without this ip-detector won't be sending the correct connectivity information when the port on broker is changed.
**How was this change tested:**
Unit tests and UAT(s)
**Any additional information or context required to review the change:**

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java [hunk 1]
```diff
@@ -61,10 +72,38 @@ public void startup() throws InterruptedException {
...
+    }
+
+    private void subscribeToMqttPortConfig() {
+        portConfig.subscribe((whatHappened, node) -> {
+            Integer port = Coerce.toInt(portConfig);
+            if (port != null) {
+                ipDetectorConfig.setMqttPort(port);
+                logger.atInfo().kv("port", port).log("MQTT broker configuration updated");
+            }
+            ipDetectorManager.updateIps(ipDetectorConfig);
```
[the remark this answers is not in the document]
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:106]: Yes, the thread can take up to 60 sec, which might not be a good customer experience

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java [hunk 2]
```diff
@@ -61,10 +72,38 @@ public void startup() throws InterruptedException {
      * @throws  InterruptedException if the thread interrupted
      */
     @Override
-    public void shutdown() throws InterruptedException {
+    public synchronized void shutdown() throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:75]: No, I am not calling this method. I saw even super methods are not synchronized. I think we should keep this so that we can avoid any unforeseen problem. Let me know your thoughts

File: src/test/java/com/aws/greengrass/detector/IpDetectorServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+        Topics mockMoquetteConfig = mock(Topics.class);
+        lenient().when(moquetteService.getConfig()).thenReturn(mockMoquetteConfig);
+
+        Topic portTopic = Topic.of(context, IpDetectorService.DETECTOR_SERVICE_NAME, TestConstants.PORT_2);
+        lenient().when(mockMoquetteConfig.find(KernelConfigResolver.CONFIGURATION_CONFIG_KEY,
+                IpDetectorService.MOQUETTE, IpDetectorService.PORT)).thenReturn(portTopic);
+
+        ipDetectorService.startup();
+        Thread.sleep(10000);
```
[the remark this answers is not in the document]
Comment (k-khatri) [reply src/test/java/com/aws/greengrass/detector/IpDetectorServiceTest.java]: Thanks it was by mistake I forgot to remove it.

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java [hunk 4]
```diff
@@ -61,10 +72,38 @@ public void startup() throws InterruptedException {
...
     }
+
+    private void subscribeToConfigs() {
+        if (portConfig == null) {
+            try {
+                // TODO: MQTT port config should be read from a variant namespace. However,
+                // for now we need to read directly from Moquette
+                portConfig = kernel.locate(MOQUETTE_SERVICE_NAME).getConfig()
+                        .find(KernelConfigResolver.CONFIGURATION_CONFIG_KEY, MOQUETTE, PORT);
+                logger.atInfo().log("Successfully loaded Moquette service configuration");
```
[the remark this answers is not in the document]
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:89]: Added the log

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (sushantAtAws) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:75]: I agree with Jason, Component lifecycle should ensure non-concurrency here.

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java [hunk 5]
```diff
@@ -5,51 +5,62 @@
...
-    public static final String DECTECTOR_SERVICE_NAME = "aws.greengrass.clientdevices.IPDetector";
+    public static final String DETECTOR_SERVICE_NAME = "aws.greengrass.clientdevices.IPDetector";
+    static final String MOQUETTE_SERVICE_NAME = "aws.greengrass.clientdevices.mqtt.Moquette";
+    static final String MOQUETTE = "moquette";
+    static final String PORT = "ssl_port";
     private final IpDetectorManager ipDetectorManager;
     private final ScheduledExecutorService scheduledExecutorService;
-    private Future<?> future;
     private final Config ipDetectorConfig;
+    private Future<?> future;
```
Comment (sushantAtAws) [inline src/main/java/com/aws/greengrass/detector/IpDetectorService.java:34]: Better variable naming.

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (sushantAtAws) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:89]: Since you have a log at line 92, this one is not required.

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (sushantAtAws) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:107]: In that case, shouldn't we move to line 105, inside the `if` clause?

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java [hunk 6]
```diff
@@ -5,51 +5,62 @@
...
 
 
-@ImplementsService(name = IpDetectorService.DECTECTOR_SERVICE_NAME)
+@ImplementsService(name = IpDetectorService.DETECTOR_SERVICE_NAME)
 public class IpDetectorService extends PluginService {
-    public static final String DECTECTOR_SERVICE_NAME = "aws.greengrass.clientdevices.IPDetector";
+    public static final String DETECTOR_SERVICE_NAME = "aws.greengrass.clientdevices.IPDetector";
+    static final String MOQUETTE_SERVICE_NAME = "aws.greengrass.clientdevices.mqtt.Moquette";
+    static final String MOQUETTE = "moquette";
+    static final String PORT = "ssl_port";
```
Comment (sushantAtAws) [inline src/main/java/com/aws/greengrass/detector/IpDetectorService.java:30]: I want to know what happens when the port=`0` provided by the customer for the ssl_port config to moquette. Usually it means to select a random port. In that case, this approach will not work and we will need moquette to publish the resolved port.

File: src/main/java/com/aws/greengrass/detector/config/Config.java [hunk 7]
```diff
@@ -30,27 +32,22 @@
...
-                                    DEFAULT_INCLUDE_IPV4_LINK_LOCAL_ADDRESSES,
-                                    INCLUDE_IPV4_LINK_LOCAL_ADDRESSES_CONFIG_KEY)));
-        });
+                this.includeIPv4LoopbackAddrs.set(Coerce.toBoolean(configurationTopics
+                        .findOrDefault(DEFAULT_INCLUDE_IPV4_LOOPBACK_ADDRESSES,
+                                INCLUDE_IPV4_LOOPBACK_ADDRESSES_CONFIG_KEY)));
+                this.includeIPv4LinkLocalAddrs.set(Coerce.toBoolean(configurationTopics
+                        .findOrDefault(DEFAULT_INCLUDE_IPV4_LINK_LOCAL_ADDRESSES,
+                                INCLUDE_IPV4_LINK_LOCAL_ADDRESSES_CONFIG_KEY)));
+            });
```
Comment (sushantAtAws) [inline src/main/java/com/aws/greengrass/detector/config/Config.java:49]: Can we add a log here for updated config?

File: src/main/java/com/aws/greengrass/detector/uploader/ConnectivityUpdater.java [hunk 8]
```diff
@@ -79,10 +82,10 @@ synchronized void uploadAddresses(List<String> ips, Config config) {
     }
 
     //Default for JUnit Testing
-    boolean hasIpsChanged(@NonNull List<String> ips) {
+    boolean hasIpsOrPortChanged(@NonNull List<String> ips, int port) {
```
Comment (sushantAtAws) [inline src/main/java/com/aws/greengrass/detector/uploader/ConnectivityUpdater.java:85]: Lets just keep a separate function for `hasPortChanged` and use `if (hasIpsChanged || hasPortChanged)`
Smaller functions with single responsibilities are preferred.

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbutler) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:89]: Did we look at the TODO? Since we're not trying to sneak this in before launch, we should just do it right.

[the remark this answers is not in the document]
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:89]: @jbutler by variant namespace we mean custom path provided by user. So will this custom path be given as a part of IPD config?

[the remark this answers is not in the document]
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:89]: @sushantAtAws 'portConfig' can be null so we do need this log. We faced this issue in our release where we were not able to conclude where the code flow is stuck.

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java (hunk 6 above)
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java:30]: That is correct. If we are supporting the port 0 as random case, we need to update the moquette code for that case. Also in all our future brokers if we are supporting this.

File: src/main/java/com/aws/greengrass/detector/config/Config.java (hunk 7 above)
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/config/Config.java:57]: Done

File: src/main/java/com/aws/greengrass/detector/uploader/ConnectivityUpdater.java (hunk 8 above)
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/uploader/ConnectivityUpdater.java]: Done

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java]: @sushantAtAws Yes, good catch

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java (hunk 5 above)
Comment (k-khatri) [reply src/main/java/com/aws/greengrass/detector/IpDetectorService.java]: Done

File: src/main/java/com/aws/greengrass/detector/IpDetectorService.java [hunk 9]
```diff
@@ -63,9 +67,29 @@ public void startup() throws InterruptedException {
...
+
+    private void subscribeToMqttPortConfig() {
+        portConfig.subscribe((whatHappened, node) -> {
+            Integer port = Coerce.toInt(portConfig);
+            if (port != null) {
+                ipDetectorConfig.setMqttPort(port);
+                ipDetectorManager.updateIps(ipDetectorConfig);
+            }
+        });
+    }
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/detector/IpDetectorService.java:94]: This file doesn't need any changes (except fixing the typo above). The config object reads config

File: src/main/java/com/aws/greengrass/detector/config/Config.java [hunk 10]
```diff
@@ -7,50 +7,55 @@
...
      *
      * @param topics Root Configuration topic
      */
     public Config(Topics topics) {
+        this.logger = LogManager.getLogger(this.getClass()).createChild();
         Topics configurationTopics = topics.lookupTopics(KernelConfigResolver.CONFIGURATION_CONFIG_KEY);
-        configurationTopics.subscribe((whatHappened, node) -> {
-            // Hardcoding port for now till MQTT Broker is not publishing it.
-            this.mqttPort = new AtomicInteger(DEFAULT_MQTT_PORT);
+        if (configurationTopics != null) {
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/detector/config/Config.java:39]: This should throw an exception if `lookupTopics` fails. This null check results in a silent failure. I'd remove this.

Same comment as above - please back out most of these changes. Nothing is needed except reading port from config, and then we can add the log Sushant asked for

File: src/main/java/com/aws/greengrass/detector/config/Config.java [hunk 11]
```diff
@@ -76,6 +81,11 @@ public boolean isIncludeIPv4LinkLocalAddrs() {
     public int getMqttPort() {
         return this.mqttPort.get();
     }
+
+    public void setMqttPort(int mqttPort) {
+        this.mqttPort.set(mqttPort);
+        logger.atInfo().kv("port", mqttPort).log("MQTT broker configuration updated");
+    }
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/detector/config/Config.java:88]: Not needed

File: src/main/java/com/aws/greengrass/detector/config/Config.java [hunk 12]
```diff
@@ -7,50 +7,55 @@
...
+import com.aws.greengrass.logging.impl.LogManager;
 import com.aws.greengrass.util.Coerce;
 
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
 
+@SuppressWarnings("PMD.DataClass")
 public class Config {
     static final String INCLUDE_IPV4_LOOPBACK_ADDRESSES_CONFIG_KEY = "includeIPv4LoopbackAddrs";
     static final String INCLUDE_IPV4_LINK_LOCAL_ADDRESSES_CONFIG_KEY = "includeIPv4LinkLocalAddrs";
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/detector/config/Config.java:20]: Let's name the new config `defaultPort`.

File: src/test/java/com/aws/greengrass/detector/IpDetectorServiceTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (jbutler) [reply src/test/java/com/aws/greengrass/detector/IpDetectorServiceTest.java:54]: We can't add 20 seconds of sleeps to unit tests.

**Review by jbutler — CHANGES_REQUESTED** [review]

File: src/test/java/com/aws/greengrass/detector/uploader/ConnectivityUpdaterTest.java [hunk 13]
```diff
@@ -108,15 +108,23 @@ public void GIVEN_ips_changed_WHEN_get_ip_addresses_THEN_return_true() {
...
     @Test
-    public void GIVEN_ips_not_changed_WHEN_get_ip_addresses_THEN_return_false() {
+    public void GIVEN_ips_and_port_not_changed_WHEN_has_ips_or_port_changed_THEN_return_false() {
         connectivityUpdater = new ConnectivityUpdater(deviceConfiguration,null);
-        connectivityUpdater.setIpAddresses(getIps());
-        assertFalse( connectivityUpdater.hasIpsChanged(getIps()));
+        connectivityUpdater.setIpAddressesAndPort(getIps(), TestConstants.PORT_1);
+        assertFalse(connectivityUpdater.hasIpsChanged(getIps()));
+        assertFalse(connectivityUpdater.hasPortChanged(TestConstants.PORT_1));
+    }
```
Comment (jbutler) [inline src/test/java/com/aws/greengrass/detector/uploader/ConnectivityUpdaterTest.java:121]: This doesn't exercise any production code - all three of these methods are unit test only. It doesn't give us any confidence that the Updater is doing what it should be doing.

This should be tested by mocking the config and clientfactory objects and making sure that the public methods are doing what they're supposed to.

File: src/main/java/com/aws/greengrass/detector/uploader/ConnectivityUpdater.java [hunk 14]
```diff
@@ -100,7 +107,8 @@ private UpdateConnectivityInfoResponse updateConnectivityInfo(List<ConnectivityI
     }
 
     //For Junit Testing
```
Comment (jbutler) [inline src/main/java/com/aws/greengrass/detector/uploader/ConnectivityUpdater.java:109]: TODO: These unit tests should be re-written to avoid needing these testing helper methods. Do this as part of a separate PR

**Review by jbutler — APPROVED** [review]

File: src/main/java/com/aws/greengrass/detector/config/Config.java [hunk 15]
```diff
@@ -7,19 +7,24 @@
...
-    static final int DEFAULT_MQTT_PORT = 8883;
+    static final String DEFAULT_PORT_CONFIG_KEY = "defaultPort";
     static final boolean DEFAULT_INCLUDE_IPV4_LOOPBACK_ADDRESSES = false;
     static final boolean DEFAULT_INCLUDE_IPV4_LINK_LOCAL_ADDRESSES = false;
+    static final int DEFAULT_PORT = 8883;
 
-    private AtomicInteger mqttPort = new AtomicInteger(DEFAULT_MQTT_PORT);
+    private AtomicInteger defaultPort = new AtomicInteger(DEFAULT_PORT);
```
Comment (wikimonkey) [inline src/main/java/com/aws/greengrass/detector/config/Config.java:27]: A side question: why do we use `AtomicXXX` here? Are we using their atomic compound operations? If we just do primitive assignment, I think it's intrinsically atomic in Java.

**Review by wikimonkey — APPROVED** [review]
